File: problems/236_lowest_common_ancestor_of_a_binary_tree/Solution.hs
-- | 236. Lowest Common Ancestor of a Binary Tree
module Solution where

import Control.Applicative ((<|>))

data Tree a
  = Empty
  | Node a (Tree a) (Tree a)
  deriving (Show, Eq)

lowestCommonAncestor :: Eq a => Tree a -> a -> a -> Maybe a
lowestCommonAncestor Empty _ _ = Nothing
lowestCommonAncestor (Node val left right) p q
  | val == p || val == q = Just val
  | otherwise =
      let l = lowestCommonAncestor left p q
          r = lowestCommonAncestor right p q
       in case (l, r) of
            (Just _, Just _) -> Just val
            _ -> l <|> r

File: problems/236_lowest_common_ancestor_of_a_binary_tree/solution.cpp
// Leetcode 236. Lowest Common Ancestor of a Binary Tree
//
// Two cases:
// (1)   p    (2)  x
//     /          / \
//    q          p   q
//
// Key insight: p and q are guaranteed to exist in the tree.
// If we hit p or q first, it must be the LCA (Case 1).
class Solution {
 public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    if (!root || root == p || root == q) {
      return root;
    }

    TreeNode* left = lowestCommonAncestor(root->left, p, q);
    TreeNode* right = lowestCommonAncestor(root->right, p, q);

    if (left && right) {  // p and q found in different subtrees
      return root;
    }

    return left ? left : right;
  }
};
